### PR TITLE
Feature/android keystore implementation

### PR DESCRIPTION
**Pull request by drilonreqica**

### CODE REVIEW

File: secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/KeystoreTool.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+                .build();
+
+        KeyPairGenerator generator
+                = KeyPairGenerator.getInstance(KEY_ENCRYPTION_ALGORITHM, KEY_KEYSTORE_NAME);
+        generator.initialize(spec);
+
+        generator.generateKeyPair();
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
```
Comment (itsmortoncornelius) [inline secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/KeystoreTool.java]: @RequiresApi(Build.VERSION_CODES.JELLY_BEAN_MR2) should be sufficient

File: secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/SecurePreferences.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+import javax.crypto.NoSuchPaddingException;
+
+import static android.content.Context.MODE_PRIVATE;
+
+public class SecurePreferences {
+
+    private static final String KEY_SHARED_PREFERENCES_NAME = "SecurePreferences";
+
+    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
```
Comment (itsmortoncornelius) [inline secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/SecurePreferences.java]: @RequiresApi(Build.VERSION_CODES.JELLY_BEAN_MR2) should be sufficient

File: secure-device-storage/sds-android/app/build.gradle [hunk 3]
```diff
@@ -21,4 +21,5 @@ android {
 dependencies {
     compile fileTree(dir: 'libs', include: ['*.jar'])
     compile 'com.android.support:appcompat-v7:25.1.1'
+    compile project(path: ':securedevicestorage')
```
Comment (drilonreqica) [inline secure-device-storage/sds-android/app/build.gradle]: Fixed in commit: https://github.com/adorsys/secure-key-storage/pull/54/commits/825b5723dd621d327ac110cc7399f8372493cca4

File: secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/KeystoreTool.java (hunk 1 above)
Comment (drilonreqica) [inline secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/KeystoreTool.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/pull/54/commits/825b5723dd621d327ac110cc7399f8372493cca4

File: secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/SecurePreferences.java (hunk 2 above)
Comment (drilonreqica) [inline secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/SecurePreferences.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/pull/54/commits/825b5723dd621d327ac110cc7399f8372493cca4

File: secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/KeystoreTool.java [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+            output = Cipher.getInstance(KEY_TRANSFORMATION_ALGORITHM, KEY_CIPHER_MARSHMALLOW_PROVIDER);
+        } else {
+            return null;
+        }
+
+        output.init(Cipher.DECRYPT_MODE, getPrivateKey(context));
+
+        CipherInputStream cipherInputStream = new CipherInputStream(
+                new ByteArrayInputStream(Base64.decode(encryptedMessage, Base64.DEFAULT)), output);
+        ArrayList<Byte> values = new ArrayList<>();
```
Comment (drilonreqica) [inline secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/KeystoreTool.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/pull/54/commits/6b0ab96771c505e7df50bfa861fdbca6c765a84c

File: secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/KeystoreTool.java [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+    static void generateKeyPair(@NonNull Context context)
+            throws InvalidAlgorithmParameterException,
+            NoSuchProviderException, NoSuchAlgorithmException, UnrecoverableKeyException,
+            CertificateException, KeyStoreException, IOException {
+
+        // Create new key if needed
+        if (!keyPairExists()) {
+            if (Build.VERSION.SDK_INT >= M) {
+                generateJellyBeanKeyPair(context);
```
Comment (drilonreqica) [inline secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/KeystoreTool.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/pull/54/commits/6b0ab96771c505e7df50bfa861fdbca6c765a84c

Comment (drilonreqica) [inline secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/KeystoreTool.java]: Update if-else clause. For now we have to use the JellyBean implementation on all API until we implement the Marshmallow logic.

**Pull request merged**